L8 product in stores

Tasenga opened the pull request:
Task 2: L8 product in stores
In 'analysis of products in stores' module:
1) Task_2.py to create file L8_result.csv containing information about:
                - number of unique products and stores;
                - the user who approved the highest number of prices;
                - the number of products sold in each store;
                - the average cost of each product;
                - shops which selling the most expensive and cheapest product (indicating the product and its price).

Commit:
created new module 'analysis of products in stores':

- files moved to other directories,
- informations about new module was included in README file

Commit:
Merge branch 'development' into L8_product_in_stores

# Conflicts:
#	.gitignore
#	Task_1.py
#	Task_3.py
#	analysis of reviews from Amazon/Task_1.py
#	analysis of reviews from Amazon/Task_2.py
#	analysis of reviews from Amazon/Task_3.py
#	analysis of reviews from Amazon/document_creation.py
#	analysis_of_reviews_from_Amazon/Task_1_1.py
#	analysis_of_reviews_from_Amazon/Task_1_3.py
#	analysis_of_reviews_from_Amazon/document_creation.py
#	document_creation.py

Commit:
remove old files

Commit:
count_unique_values - function returns the number of unique values for a field which was given like function's parameter

count_unique_values - function returns the number of unique values for a field which was given like function's parameter
get_user_by_approves - function returns the user name who approved the highest number of prices
get_products_in_stores - function returns the number of products sold in each store

Commit:
added module 'black'

Commit:
get_avg_cost_per_products - function returns the average cost of each product

Commit:
create new folder for common functions for all project modules (add common function 'save_file')

add function 'get_exp_and_chp_products' - function returns the shops which selling the most expensive.

Commit:
readme updated

sleshJdev commented:
Please, resolve conflicts

sleshJdev commented on analysis_of_products_in_stores/Task_2_1.py:
```diff
@@ -0,0 +1,142 @@
+    with open(os.path.join(ROOT_DIR, filename)) as file:
+        reader = csv.reader(file)
+        next(reader)
+        main_dict = [
```
It's not dict, it's list

Tasenga replied:
renamed it like main_data

sleshJdev commented on analysis_of_products_in_stores/Task_2_1.py:
```diff
@@ -0,0 +1,142 @@
+def get_data(filename):
+    """function returns a dictionary from the datafile"""
+
+    ROOT_DIR = os.path.dirname(os.path.abspath(__file__))
```
Please, import dirname/abspath directly to save some space

Tasenga replied:
done: from os.path import dirname, abspath, join

sleshJdev commented on analysis_of_products_in_stores/Task_2_1.py:
```diff
@@ -0,0 +1,142 @@
+    with open(os.path.join(ROOT_DIR, filename)) as file:
+        reader = csv.reader(file)
+        next(reader)
+        main_dict = [
```
Let's do return directly from here. Please [check](https://www.python.org/dev/peps/pep-0343/#id46)

Tasenga replied:
done

`def get_data(filename):
    """function returns a dictionary from the datafile"""

    ROOT_DIR = dirname(abspath(__file__))
    with open(join(ROOT_DIR, filename)) as file:
        reader = csv.reader(file)
        next(reader)
        main_data = [
            {
                "ITEM_ID": row[0],
                "STORE_ID": row[1],
                "PRICE": float(row[2]),
                "APPROVED_BY": row[3],
            }
            for row in reader
        ]
        return main_data`

sleshJdev replied:
Why not:
```
return [
        {
            "ITEM_ID": row[0],
            "STORE_ID": row[1],
            "PRICE": float(row[2]),
            "APPROVED_BY": row[3],
        }
        for row in reader
    ]
``` ?

Tasenga replied:
Without reason. I didn't understand that it is possible after reading documentation. My fault, fixed it.

sleshJdev commented on analysis_of_products_in_stores/Task_2_1.py:
```diff
@@ -0,0 +1,142 @@
+    }
+    return grouped_dict
+
+def count_unique_values(dict):
```
len(dict)?

Tasenga replied:
oh, exactly. Done.

sleshJdev replied:
Let's get rid of this function at all, it doesn't seem to be useful, because it just call `len(dict)`. Creating a function for the sake of creating the function is doubtful.

Tasenga replied:
Done

sleshJdev commented on analysis_of_products_in_stores/Task_2_1.py:
```diff
@@ -0,0 +1,142 @@
+def sort_dict(dict, field):
+    """function returns a sorted dictionary for a field which was given like function's parameter and this parameter"""
+
+    sorted_dict = sorted(dict, key=lambda position: position[field])
```
Is this function re-used somewhere? It seems a bit useless as it just calls built-in sorted func. For the caller of this function no difference whether it calls sort_dict or sorted, it's almost the same. Also, you will need to deal with tuples, which is also a kind of inconvenience. So lets inline this function.

Tasenga replied:
included it in 'group_dict' function

`def group_dict(dict, field):
    """function returns a grouped dictionary"""
    sorted_dict = sorted(dict, key=lambda position: position[field])
    grouped_dict = {
        key: [group for group in groups]
        for key, groups
        in groupby(sorted_dict[0], key=lambda position: position[sorted_dict[1]])
    }
    return grouped_dict`

sleshJdev replied:
just fyi: to insert code block we have to use triple apostrophe sign. Please, check [this](https://help.github.com/en/github/writing-on-github/creating-and-highlighting-code-blocks#fenced-code-blocks)

Tasenga replied:
okay, I will try it. I used "insert code" button on the toolbar.

sleshJdev replied:
It can be even simple:
```
def group_dict(dict, field):
    return groupby(
        sorted(dict, key=lambda position: position[field]),
        key=lambda position: position[field],
    )
```
or inlined at all.

In your case `key: [group for group in groups]` is not neede, because gorupby returns you `{key -> [items with the key]}` dict.

sleshJdev commented on analysis_of_products_in_stores/Task_2_1.py:
```diff
@@ -0,0 +1,142 @@
+def group_dict(def_sort_dict):
+    """function returns a grouped dictionary"""
+
+    grouped_dict = {
```
The same as for sort_dict, if it's used once, let's inline it.

Commit:
Merge branch 'L2_git_hooks' into L8_product_in_stores

# Conflicts:
#	git.hooks/pre-push

Tasenga commented:
> 
> 
> Please, resolve conflicts

resolved

sleshJdev commented on analysis_of_products_in_stores/Task_2_1.py:
```diff
@@ -0,0 +1,142 @@
+def get_user_by_approves(dict):
+    """function returns the user name who approved the highest number of prices"""
+
+    result = max(((key, len(dict[key])) for key in dict.keys()), key=lambda x: x[1])
```
It may be a be efficient:
```
key, items = max(d.items(), key=lambda t: len(t[1]))
return (key, len(items))
```

sleshJdev replied:
What about this?

sleshJdev commented on analysis_of_products_in_stores/Task_2_1.py:
```diff
@@ -0,0 +1,142 @@
+    """function returns the number of products sold in each store"""
+
+    result = {
+        STORE_ID: len(set(position["ITEM_ID"] for position in positions))
```
Why upper case?

Tasenga replied:
so as upper case uses in incoming data like name of field. I use same to save the same style

sleshJdev replied:
doubtfully, but ok

sleshJdev commented on common_functions/work_with_document.py:
```diff
@@ -0,0 +1,20 @@
+def create_directory():
+    '''function creates a directory to save files'''
+    try:
+        os.mkdir(os.path.join(os.environ['PYTHONPATH'], 'resulting data'))
```
We cannot rely on `os.environ['PYTHONPATH']`, for example. here is my value of PYTHONPATH: 
```
/home/yauheni/work/bigdatalab/2020/Tasenga/ITechArt_Android_App_Review:/home/yauheni/.IntelliJIdea2019.3/config/plugins/python/helpers/pycharm_matplotlib_backend:/home/yauheni/.IntelliJIdea2019.3/config/plugins/python/helpers/pycharm_display:/home/yauheni/.IntelliJIdea2019.3/config/plugins/python/helpers/third_party/thriftpy:/home/yauheni/.IntelliJIdea2019.3/config/plugins/python/helpers/pydev:/home/yauheni/.IntelliJIdea2019.3/system/cythonExtensions:/home/yauheni/work/bigdatalab/2020/Tasenga/ITechArt_Android_App_Review/analysis_of_products_in_stores
```

sleshJdev replied:
Just let the user specify where to store results, it's much esier

sleshJdev replied:
+ Let's start using https://docs.python.org/3/library/pathlib.html#pathlib.Path.mkdir, there is exists_ok flag, then you don't need to use try-except

Tasenga replied:
Done and I added possibility to not enter the path, in this case base_path will be like path of running file

`print('Enter path to directory to save results')
    user_path = input()
    if user_path == '':
        path = dirname(abspath(__file__))
    else:
        path = user_path`

`def save_file(path, name, data, mode='w'):
    '''function creates a file from the transferred data'''
    
    Path(join(path, 'resulting data')).mkdir(parents=True, exist_ok=True)
    with open(join(path, 'resulting data', name), mode, newline='') as csv_file:`

sleshJdev replied:
It can be more idiomatic:
```
    out_dir = Path(join(path, "resulting data"))
    out_dir.mkdir(parents=True, exist_ok=True)
    with Path(out_dir, name).open(mode=mode, newline="") as csv_file:
        writer = csv.writer(csv_file, delimiter="\t")
        for line in data:
            writer.writerow(line)
        writer.writerow("")
```

Please, check [doc](https://docs.python.org/3/library/pathlib.html#pathlib.Path.open)

Tasenga replied:
I see. Done.

sleshJdev commented on git.hooks code/commit-msg:
```diff
@@ -1,5 +1,7 @@
 #!/bin/sh
 
+popen('chmod +x commit_msg', shell=True, stdout=PIPE)
```
To get this code line executed `commit_msg` should have +x flag, but if it will have this flag, then you don't need this do it by yourself.

Tasenga replied:
this string is absent in master branch, I deleted this file from current branch to save the file in master after merge

sleshJdev commented on git.hooks code/pre-push:
```diff
@@ -5,6 +5,8 @@ import sys
 from subprocess import Popen, PIPE
 import py_compile
 
+Popen('chmod +x pre-push', shell=True, stdout=PIPE)
```
The same as above

Tasenga replied:
this string is absent in master branch, I deleted this file from current branch to save the file in master after merge

sleshJdev commented on analysis_of_products_in_stores/Task_2_1.py:
```diff
@@ -0,0 +1,142 @@
+
+    save_file(
+        filename,
+        [[key, value] for key, value in get_products_in_stores(dict_by_STORE_ID).items()],
```
We unpack result of get_products_in_stores and then pack it again. It's not efficient and not needed.

Tasenga replied:
deleted not efficient part

Commit:
main_dict renamed to main_data;

'sorted' function was implemented in 'grouped' function;
function 'count_unique_values' was optimized with using 'len(dict)';
added option for user to input path to save result

Commit:
function save_result was optimized with using Path.mkdir,

added incoming parameter path, which requested from user

sleshJdev commented on requirements.txt:
```diff
@@ -0,0 +1,9 @@
+attrs==19.3.0
+black==19.10b0
+Click==7.0
+pathlib==1.0.1
```
pathlib is part of python SDK

sleshJdev commented on requirements.txt:
```diff
@@ -0,0 +1,9 @@
+appdirs==1.4.3
+attrs==19.3.0
```
Could you please specify why do you need all these dependencies?

Tasenga replied:
via 'black'

sleshJdev commented:
In the context of this task, entries with price = 0 are valid data?

Commit:
cleared requirements

Commit:
function 'save_file' was rewritten more idiomatic

Commit:
function 'get_data' was rewritten more idiomatic (return [row for row in reader] instead of return data)

function 'count_unique_values' was deleted cause useless
